contributors/design-proposals/live-and-inplace-vertical-scaling.md

```diff
@@ -0,0 +1,325 @@
+
+This section describes briefly API changes for pod-level vertical scaling and the related changes on the 'API server', the 'Scheduler', and the 'Kubelet'.
+
+* **ResizeRequest**
+
+A new data structure, ResizeRequest, is added to v1.PodSpec:
+
+```go
+const (
+    ResizeRequested ResizeStatus = "Requested"
+    ResizeAccepted ResizeStatus = “Accepted”
+    ResizeRejected ResizeStatus = “Rejected”
+    ResizeNone      ResizeStatus = "None"
+)
+
+type ResizeRequest struct {
+    RequestStatus ResizeStatus
+    NewResources  []ResourceRequirements // indexed by containers’ index
+}
+
+Type PodSpec {
+    …
+    ResizeRequest ResizeRequest
+    ...
+}
+```
+
+ResizeRequest has two variables, RequestStatus and NewResources.
+RequestStatus represents the status of a resource resizing request.
+ResizeRequested indicates resource resizing for a pod is requested.
+ResizedAccepted and ResizedRejected means that the requested resource resizing is accepted and rejected, respectively, by the Scheduler.
+The NewResources is an array indexed by a container’s index and its each entry holds new resource requirements of a container that needs to resize.
+
+Given a new PodSpec with new resource requirements from a client, first the 'API server' validates it.
+If it is valid, the 'API server' sets the RequestStatus to ResizedRequested and copies the new resource requirements of each container into the NewResources.
+Also, the 'API server' restores the resource requirements of each container of the PodSpec to the original and writes the revised PodSpec to ETCD to communicate with the Scheduler.
+This is because at this moment the PodSpec on ETCD shouldn’t be updated with new resource requirements.
+
+For a pod with ResizeRequested, the 'Scheduler' checks if the node on which the pod currently runs has enough resources to resize the pod.
+The Scheduler notify the 'API server' of the result via 'Resizing' API operation, which will be describe below.
+
+* **Resizing**
+
+A new API, Resizing, for the 'scheduler' is introduced:
+
+```go
+// Resizing resizes the resources allocated to a pod
+type Resizing struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+    Request ResizeRequest
+}
+```
+
+Resizing has the metadata of a pod to resize and a value of ResizeRequest that holds the status of a resizing request, which indicates whether the resizing is feasible or not, and new resource requirements of the pod.
+
+Once the 'Scheduler' determine whether a resource resizing on a pod is feasible, or not, it notifies to the API server via this Resizing API.
+ObjectMeta.Namespace/Name are set to the Namespace and Name of the pod to resize and ResizeRequest is set with new resource requirements.
+
+Given a Resizing API operation, the ResizeStatus of the PodSpec of a Pod is updated according to that of the Resizing operation.
+If the ResizeStatus is ResizeAccepted, the API server updates the ResourceRequirement of each container of a pod with new resource requirements on ETCD.
```

Comment:
I think the scheduler alone is insufficient.  I think both the kubelet and the scheduler need to agree that a change is possible.

Reply:
yes, both kubelet and scheduler need to agree this change.

Reply:
Is it OK that kubelet alone makes the decision? After the decision, schedule will be noticed the decision. It is a little same as that we specify 'NodeName' in pod spec when creating the spec, all the decision is made by kubelet.

Reply:
Now, I see that even when the scheduler approves a resizing request, for some reasons, the kubelet may reject the resizing. So, now I'm revising the design proposal and prototype implementation for both of the kubelet and the scheduler to make the agreement, including the modification to update the resource requirement fields in pod spec after the kubelet acknowledges the resizing as @derekwaynecarr commented.